src/highlight/highlightUtil.ts

```diff
@@ -58,3 +58,65 @@ export const centerHighlight = (shapes: Shape[]): Position => {
 export function isHighlight(annotation: Annotation): annotation is AnnotationHighlight {
     return annotation?.target?.type === Type.highlight;
 }
+
+export const dedupeRects = (rects: Shape[], threshold = 0.1): Shape[] => {
+    const dedupedRects: Shape[] = [];
+
+    rects.forEach(curr => {
+        const prev = dedupedRects.pop();
+        // empty list, push current
+        if (!prev) {
+            dedupedRects.push(curr);
+            return;
+        }
+
+        const heightThreshold = Math.max(
+            Math.floor(prev.height * threshold + 0.5),
+            Math.floor(curr.height * threshold + 0.5),
+            1,
+        );
+
+        // the same rect, push the larger one
+        if (prev.x === curr.x && prev.width === curr.width && Math.abs(prev.y - curr.y) <= heightThreshold) {
+            dedupedRects.push(prev.height > curr.height ? prev : curr);
+            return;
+        }
+
+        // different rects, push both
+        dedupedRects.push(prev);
+        dedupedRects.push(curr);
+    });
+
+    return dedupedRects;
+};
+
+export const groupByRow = (shapes: Shape[]): Record<number, Shape[]> => {
+    const rows: Record<number, Shape[]> = {};
+    shapes.forEach(shape => {
+        const { y } = shape;
+        const roundedY = Math.floor(y + 0.5);
+        if (!rows[roundedY]) {
+            rows[roundedY] = [shape];
+        } else {
+            rows[roundedY].push(shape);
+        }
+    });
+
+    return rows;
+};
+
+export const combineRectsByRow = (allShapes: Shape[]): Shape[] => {
+    const dedupedRects = dedupeRects(allShapes);
+    const rowMap = groupByRow(dedupedRects);
+    return Object.values(rowMap).map(shapes => getBoundingRect(shapes));
+};
+
+export const getShapeRelativeToPage = (
+    { height, width, x, y }: Shape,
```

Comment:
Since these are both Shape types, it's not clear what each is supposed to represent. Since it's also broken the function signature across two lines, consider naming the parameters and destructuring them in the function body.